src/libopensc/card-isoApplet.c

```diff
@@ -1218,8 +1218,13 @@ isoApplet_compute_signature(struct sc_card *card,
 		// For RSA-PSS signature schemes the IsoApplet expects only the hash.
 		u8 tmp[64]; // large enough for SHA512
 		size_t tmplen = sizeof(tmp);
-		sc_pkcs1_strip_digest_info_prefix(NULL, data, datalen, tmp, &tmplen);
-		r = iso_ops->compute_signature(card, tmp, tmplen, out, outlen);
+		r = sc_pkcs1_strip_digest_info_prefix(NULL, data, datalen, tmp, &tmplen);
+		if (r == SC_SUCCESS) {
+			r = iso_ops->compute_signature(card, tmp, tmplen, out, outlen);
+		} else {
+			/* No digest info present? Use the value as it is */
+			r = iso_ops->compute_signature(card, data, datalen, out, outlen);
+		}
```

Comment:
I think its better to return an error if the digest info is missing.

Reply:
There is no digestinfo in case the `RSA_PKCS_PSS` mechanism is used (as visible in the modified p11test results). If you use the `SHA*_RSA_PKCS_PSS` mechanism, the input is either hashed inside of the card or in OpenSC, depending on the card capabilities, but with `RSA_PKCS_PSS`, the OpenSC gets the digest only with the length of `hLen` so there is no digestinfo involved (the digest is part of the PSS_PARAMS structure though):
https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061139

At least that is how I covered this mechanism in the p11test and how it works with other cards so far. If you reading of the specs is different, feel free to correct me.

Reply:
I'm reading the spec the same way, but I still do not understand how this has worked before where the return code of `sc_pkcs1_strip_digest_info_prefix` was not checked. Shouldn't the `RSA_PKCS_PSS` test cases in p11test have failed before that change?

Reply:
I think they were failing, but given the vast amount of the combination for the RSA-PSS and that really no driver supports everything, the failures are silently ignored ...

Reply:
Understood. Forwarding the data as-is to the card if no digest is present is the correct way then :+1: